hooks/tk-multi-publish2/photoshopcc.basic/upload_version.py

```diff
@@ -0,0 +1,348 @@
+        :param item: Item to process
+
+        :returns: dictionary with boolean keys accepted, required and enabled
+        """
+
+        document = item.properties.get("document")
+        if not document:
+            self.logger.warn("Could not determine the document for item")
+            return {"accepted": False}
+
+        path = _document_path(document)
+
+        if not path:
+            # the document has not been saved before (no path determined).
+            # provide a save button. the document will need to be saved before
+            # validation will succeed.
+            self.logger.warn(
+                "The Photoshop document '%s' has not been saved." %
+                (document.name,),
+                extra=_get_save_action(document)
+            )
+
+        self.logger.info(
+            "Photoshop '%s' plugin accepted document: %s" %
+            (self.name, document.name)
+        )
+        return {
+            "accepted": True,
+            "checked": True
+        }
+
+    def validate(self, settings, item):
+        """
+        Validates the given item to check that it is ok to publish.
+
+        Returns a boolean to indicate validity.
+
+        :param settings: Dictionary of Settings. The keys are strings, matching
+            the keys returned in the settings property. The values are `Setting`
+            instances.
+        :param item: Item to process
+
+        :returns: True if item is valid, False otherwise.
+        """
+
+        document = item.properties["document"]
+        path = _document_path(document)
+
+        if not path:
+            # the document still requires saving. provide a save button.
+            # validation fails.
+            self.logger.error(
+                "The Photoshop document '%s' has not been saved." %
+                (document.name,),
+                extra=self._get_save_as_action(document)
+            )
+            return False
+
+        return True
+
+    def publish(self, settings, item):
+        """
+        Executes the publish logic for the given item and settings.
+
+        :param settings: Dictionary of Settings. The keys are strings, matching
+            the keys returned in the settings property. The values are `Setting`
+            instances.
+        :param item: Item to process
+        """
+
+        publisher = self.parent
+        engine = publisher.engine
+        document = item.properties["document"]
+
+        path = _document_path(document)
+        upload_path = path
+
+        file_info = publisher.util.get_file_path_components(path)
+        if file_info["extension"] in ["psd", "psb"]:
+
+            with engine.context_changes_disabled():
+
+                # remember the active document so that we can restore it.
+                previous_active_document = engine.adobe.app.activeDocument
+
+                # make the document being processed the active document
+                engine.adobe.app.activeDocument = document
```

Comment:
If anything between here and line 215 raises, the user's previous active document won't be reset. I'd probably either write a context manager to do it, or use a try/finally.